main/main.c

```diff
@@ -1,67 +1,240 @@
-        esp_restart();
-    }
-}
-
-// Main application
-void app_main(void)
-{
-    TaskHandle_t main_task_handle = NULL;
-    int x_task_returned = xTaskCreate(&main_task, "main_task", 2048, NULL, 4, &main_task_handle);
-
-    if (x_task_returned != pdPASS)
-    {
-        ESP_LOGE(TAG_BME280, "Error with creating the task. Returned with error code: %d", x_task_returned);
-        vTaskDelete(main_task_handle);
-        esp_restart();
-    }
-
-    for (;;)
-    {
-        vTaskDelay(pdMS_TO_TICKS(1000));
-        ESP_LOGI(TAG_MAIN, "Free heap: %" PRIu32 "", esp_get_free_heap_size());
-    }
-}
\ No newline at end of file
+// main.c
+
+#include <stdio.h>
+#include <string.h>
+#include "esp_log.h"
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+#include "bme280.h"
+#include "lora.h"
+#include "esp_wifi.h"
+#include "esp_system.h"
+#include "nvs_flash.h"
+#include "esp_event.h"
+#include "esp_netif.h"
+#include "mqtt_client.h"
+#include "protocol_examples_common.h"
+#include "low_power_mode.h"
+
+#define TAG "Main"
+
+#if CONFIG_LORA_TRANSMITTER
+static void task_tx(void *pvParameters);
+
+static void initialize_sensors(void)
+{
+    esp_err_t bme_rc = ESP_OK;
+    bme_rc += bme280_init_driver(CONFIG_BME280_I2C_ADDRESS);
+    bme_rc += bme280_set_oversamp(BME280_OVERSAMP_16X, BME280_OVERSAMP_16X, BME280_OVERSAMP_16X);
+    bme_rc += bme280_set_settings(STANDBY_10MS, BME280_FILTER_COEFF_16, BME280_NORMAL_MODE);
+
+    if (ESP_OK != bme_rc)
+    {
+        ESP_LOGE("BME", "BME280 settings failed");
+        // not sure if this should restart, but also not sure what else to do here?
+        esp_restart();
+    }
+}
+#endif
+
+#if CONFIG_LORA_RECEIVER
+static void task_rx(void *pvParameters);
+#endif
+
+#if CONFIG_LORA_RECEIVER && CONFIG_ENABLE_MQTT
+static const char *MQTT_TAG = "MQTT";
+static esp_mqtt_client_handle_t mqtt_client;
+
+static void initialize_mqtt(void)
+{
+    ESP_ERROR_CHECK(nvs_flash_init());
+    ESP_ERROR_CHECK(esp_netif_init());
+    ESP_ERROR_CHECK(esp_event_loop_create_default());
+    ESP_ERROR_CHECK(example_connect());
+
+    esp_mqtt_client_config_t mqtt_cfg = {
+        .broker.address.hostname = "158.180.60.2",
+        .broker.address.port = 1883,
+        .broker.address.transport = MQTT_TRANSPORT_OVER_TCP,
+        .credentials.username = "admin",
+        .credentials.authentication.password = "",
+    };
+
+    mqtt_client = esp_mqtt_client_init(&mqtt_cfg);
+    esp_mqtt_client_start(mqtt_client);
+}
+#endif
+
+void app_main(void)
+{
+    ESP_LOGI("MAIN", "Initializing LoRa");
+    if (LORA_OK != lora_init())
+    {
+        ESP_LOGE("MAIN", "LoRa initialization failed");
+        // not sure if this should restart, but also not sure what else to do here?
+        esp_restart();
+    }
+
+#if CONFIG_LORA_TRANSMITTER
+    initialize_sensors();
+    xTaskCreate(&task_tx, "TX", 1024 * 3, NULL, 5, NULL);
+#endif
+
+#if CONFIG_LORA_RECEIVER && CONFIG_ENABLE_MQTT
+    initialize_mqtt();

```

Comment:
How do we know if MQTT initialization success and we can operate on it?

Reply:
we will make a nicer implementation at a later date, well put it in a separate component